### DISCUSSION TITLE
Refresh expired bearer token in drm access

### DISCUSSION THREAD

**Discussion by danieldulcettiBN** [TheWidlarzGroup/react-native-video#4952] [Ideas]
### What problem are you trying to solve?

I need to refresh my url (streaming video) periodically since it has a token that expires after some time.

I have already managed this flow (I ask the new url before it expires) but screen appears blank for about a second after url refreshing, probably caused by source reloading.

I also tried putting the token in the source header, but nothing.

### Proposed solution

The source must be changed without the user noticing the transition. Changing the source (same url with different token) while it plays should swap the source seamlessly, so the user does not see a blank screen during a DRM/bearer token refresh.

### Alternatives considered

I also tried putting the token in the source header, but nothing.

### Platforms

Android TV, Apple tvOS

### Additional context

**Reproduction steps**
1. Change source (same url with different token) while it plays

**Reported version:** 5.2.0

Thanks to everyone!

**Comment by freeboub** [comment]
It is expected that changing the url change playback.
For sure, you should use header to achieve that, but I don't know how to do it or if it is really possible...
Notice that as nobody can reproduce the issue as your server api is not open, you will have to investigate it

**Comment by falk-stefan** [comment]
@freeboub How can we refresh an access token if it expires mid-stream?

Currently I am providing the token by setting `headers` via `source` like this: 

```javascript
source={{
  uri: trackUri,
  headers: {
    Authorization: `Bearer ${accessToken}`,
  },
}}
```

But how can we refresh it on expiry? :thinking:

**Comment by freeboub** [comment]
@falk-stefan unfortunately I don't have any solution for now ... I will reopen the ticket to track it, but as I don't have any way to reproduce the issue, I cannot investigate... 
In terms of api I can do some suggestions if needed.

**Comment by falk-stefan** [comment]
@freeboub I have a different requirement though- I do not need to refresh the URL - but the `Authorization` header once a request reteurns HTTP 401 Token Expired. 

It's not the exact same as what @danieldulcettiBN tries to do.

**Comment by freeboub** [comment]
Yes, I agree but it is more or less the same use case. 
The change would be to add a callback onAuthError. When this callback is called application would need to call another function api to refresh bearer token (or the url)

**Comment by falk-stefan** [comment]
Hm, a more generic way would be to just return the full response on error and let the user do whatever is right. Perhaps the onAuthError could be a convenience thing but generally speaking it would probably be nice to have a general way to either intercept requests and/or responses and open a path to retry a request if needed.

**Comment by freeboub** [comment]
You mean the onError callback should be enough. I think it makes sense. Another ticket exists about error code clean up, they should be linked then.
Something I am not sure about and should be experienced, about the timing, would JS be fast enough to renew a token fastly enough and provide to react native video to get a seemless renew ...

**Comment by falk-stefan** [comment]
> You mean the onError callback should be enough.

Alsmost. I'd probably call it `onRequestError(url: string, request: RequestInit, response: Response): Promise<Response>` - this way a caller can implement some retry-logic e.g. in order to refresh the token:

```typescript
<Video onRequestError={(url, req: , res) => {
    if (res.status === 401) {
      const apiError = await res.json() as ApiError;
      if (apiError.code === 'auth/token-expired') {
        // Refresh the token
        const newToken = refreshToken();
        req.headers = {
          ...req.headers,
          Authorization: `Bearer ${newToken}`,  // Replace old header
        };
        return await fetch(url, req);
      }
    }
    return res;
  }} />
```

**Comment by freeboub** [comment]
Yes I was thinking of something similar!

**Comment by falk-stefan** [comment]
Something like this would be awesome. :)

**Comment by zoriya** [comment]
We might want to change headers without it being an error, for example if the token comes with an expiry date (common with JWTs).

For example, hls.js gives a xhrSetup async method where you can edit the requests before they are sent.

I think just being able to change headers without resetting the source would be enough.

Users could then combine it with a custom onError handler that would refresh the token and retry if it got a 401/403 if they need to.

**Comment by freeboub** [comment]
For android, the good implementation to do is detailed here: https://developer.android.com/media/media3/exoplayer/customization
See "Customizing server interactions" section, and "ResolvingDataSource" usage

**Comment by TWG-AI** [comment]
Thanks for the idea — noted for the team to consider.

> I'm an AI assistant, so double-check any specifics I may have mentioned.
